packages/controls/src/controls/shape-v2/shape-v2.ts

```diff
@@ -0,0 +1,322 @@
+import { z } from 'zod'
+
+import { mapValues } from '../../lib/functional'
+import { safeParse, type ParseResult } from '../../lib/zod'
+
+import { ControlDataTypeKey, type Data } from '../../common'
+import {
+  type CopyContext,
+  type MergeTranslatableDataContext,
+} from '../../context'
+import { type IntrospectionTarget } from '../../introspection'
+import {
+  SerializationSchema,
+  type DeserializedRecord,
+  type SerializedRecord,
+} from '../../serialization'
+
+import {
+  type DataType as DataType_,
+  type ResolvedValueType as ResolvedValueType_,
+  type ValueType as ValueType_,
+} from '../associated-types'
+import { ControlDefinition, serialize, type SchemaType } from '../definition'
+import { type SendMessage } from '../instance'
+
+import { ShapeV2Control } from './shape-v2-control'
+
+type ItemDefinition = ControlDefinition<string, unknown, any, any, any>
+type KeyDefinitions = Record<string, ItemDefinition>
+
+export type Config<Defs extends KeyDefinitions = KeyDefinitions> = {
+  label?: string
+  layout?: typeof Definition.Layout.Inline | typeof Definition.Layout.Popover
+  readonly type: Defs
+}
+
+export type ShapeV2DataTypeV0<C extends Config> = {
+  [K in keyof C['type']]?: DataType_<C['type'][K]>
+}
```

Comment:
Do we need to define/publicly expose this? I see that you are using it in the builder, but IMHO it'd be clearer/more precise to use `DataType<ShapeDefinition<C>>` for the v0 data.

Reply:
> IMHO it'd be clearer/more precise to use DataType<ShapeDefinition<C>> for the v0 data.
Nice. I agree.

What are your thoughts on `ShapeV2DataTypeV1`?

Reply:
Unfortunately, there is no way to scope types in TypeScript. We did end up exposing `RichTextDataV2` for rich text v2.

Reply:
With the export new static `versionData` function I no longer need this in cosmos.